src/scripts/http/cachedHttp.ts

```diff
@@ -0,0 +1,121 @@
+import {ResponsePackage} from "../responsePackage";
+
+import {Storage} from "../storage/storage";
+
+export interface TimeStampedData {
+	data: any;
+	lastUpdated: number;
+}
+
+export type GetResponseAsync = () => Promise<ResponsePackage<string>>;
+
+/**
+ * Allows the creation of HTTP GET requests to retrieve data, as well as caching of the result.
+ */
+export class CachedHttp {
+	private cache: Storage;
+
+	constructor(cache: Storage) {
+		this.cache = cache;
+	}
+
+	/**
+	 * Given a key, checks the cache for a fresh copy of that value. If the cached value does
+	 * not exist, or is not fresh, fetches a fresh copy of the data from the specified endpoint.
+	 */
+	public getAndCacheFreshValue(key: string, getRemoteValue: GetResponseAsync, updateInterval: number): Promise<TimeStampedData> {
+		if (!key) {
+			throw new Error("key must be a non-empty string, but was: " + key);
+		} else if (!getRemoteValue) {
+			throw new Error("getRemoteValue must be non-undefined");
+		} else if (updateInterval < 0) {
+			updateInterval = 0;
+		}
+
+		return new Promise<TimeStampedData>((resolve, reject) => {
+			let value = this.cache.getValue(key);
+			let keyIsPresent = !!value;
+			if (keyIsPresent) {
+				let valueAsJson: TimeStampedData;
+				try {
+					valueAsJson = JSON.parse(value);
+				} catch (e) {
+					reject({ error: e });
+				}
+
+				let valueHasExpired = CachedHttp.valueHasExpired(valueAsJson, updateInterval);
+
+				if (!valueHasExpired) {
+					resolve(valueAsJson);
+					return;
+				}
+			}
+
+			this.getAndCacheRemoteValue(key, getRemoteValue).then((timeStampedData) => {
+				resolve(timeStampedData);
+			}, (error) => {
+				reject(error);
+			});
+		});
+	}
+
+	public getAndCacheRemoteValue(key: string, getRemoteValue: GetResponseAsync): Promise<TimeStampedData> {
```

Comment:
I'm not sure if we call this explicitly from the outside, but I do believe there are times where we actually force a fresh get of remote values (by setting updateInterval to 0). I may have forgotten to convert those calls to this new method call, so I'll do a double check :)